client/asset/zec/sign.go

```diff
@@ -0,0 +1,256 @@
+	var groupID uint32 = versionSaplingGroupID
+	if tx.Version == dexzec.VersionCanopy {
+		groupID = versionNU5GroupID
+	}
+	var nVersion [4]byte
+	binary.LittleEndian.PutUint32(nVersion[:], groupID)
+	sigHash.Write(nVersion[:])
+
+	var zeroHash chainhash.Hash
+
+	// hashPrevouts
+	if hashType&txscript.SigHashAnyOneCanPay == 0 {
+		sigHash.Write(sigHashes.HashPrevOuts[:])
+	} else {
+		sigHash.Write(zeroHash[:])
+	}
+
+	// hashSequence
+	if hashType&txscript.SigHashAnyOneCanPay == 0 &&
+		hashType&sigHashMask != txscript.SigHashSingle &&
+		hashType&sigHashMask != txscript.SigHashNone {
+
+		sigHash.Write(sigHashes.HashSequence[:])
+	} else {
+		sigHash.Write(zeroHash[:])
+	}
+
+	// hashOutputs
+	if hashType&sigHashMask != txscript.SigHashSingle && hashType&sigHashMask != txscript.SigHashNone {
+		sigHash.Write(sigHashes.HashOutputs[:])
+	} else if hashType&sigHashMask == txscript.SigHashSingle && idx < len(tx.TxOut) {
+		var (
+			b bytes.Buffer
+			h chainhash.Hash
+		)
+		if err = wire.WriteTxOut(&b, 0, 0, tx.TxOut[idx]); err != nil {
+			return nil, nil, err
+		}
+
+		if h, err = blake2bHash(b.Bytes(), []byte(outputsHashPersonalization)); err != nil {
+			return nil, nil, err
+		}
+		sigHash.Write(h.CloneBytes())
+	} else {
+		sigHash.Write(zeroHash[:])
+	}
+
+	// hashJoinSplits
+	sigHash.Write(zeroHash[:])
+
+	// hashShieldedSpends
+	if tx.Version == dexzec.VersionSapling {
+		sigHash.Write(zeroHash[:])
+	}
+
+	// hashShieldedOutputs
+	if tx.Version == dexzec.VersionSapling {
+		sigHash.Write(zeroHash[:])
+	}
+
+	// nLockTime
+	var lockTime [4]byte
+	binary.LittleEndian.PutUint32(lockTime[:], tx.LockTime)
+	sigHash.Write(lockTime[:])
+
+	// nExpiryHeight
+	var expiryTime [4]byte
+	binary.LittleEndian.PutUint32(expiryTime[:], tx.ExpiryHeight)
+	sigHash.Write(expiryTime[:])
+
+	// valueBalance
+	if tx.Version >= dexzec.VersionSapling {
+		var valueBalance [8]byte
+		binary.LittleEndian.PutUint64(valueBalance[:], 0)
+		sigHash.Write(valueBalance[:])
+	}
+
+	// hash type
+	var bHashType [4]byte
+	binary.LittleEndian.PutUint32(bHashType[:], uint32(hashType))
+	sigHash.Write(bHashType[:])
+
+	if idx != math.MaxUint32 {
+		// outpoint
+		sigHash.Write(tx.TxIn[idx].PreviousOutPoint.Hash[:])
+		var bIndex [4]byte
+		binary.LittleEndian.PutUint32(bIndex[:], tx.TxIn[idx].PreviousOutPoint.Index)
+		sigHash.Write(bIndex[:])
+
+		// scriptCode
+		if err = wire.WriteVarBytes(&sigHash, 0, subScript); err != nil {
+			return nil, nil, err
+		}
+
+		// value
+		if err = binary.Write(&sigHash, binary.LittleEndian, amt); err != nil {
+			return nil, nil, err
+		}
+
+		// nSequence
+		var bSequence [4]byte
+		binary.LittleEndian.PutUint32(bSequence[:], tx.TxIn[idx].Sequence)
+		sigHash.Write(bSequence[:])
+	}
+
+	var h chainhash.Hash
+	preImage := sigHash.Bytes()
+	sigHashKey := append([]byte(blake2BSigHash), cver[:]...)
+	if h, err = blake2bHash(preImage, sigHashKey); err != nil {
+		return nil, nil, err
+	}
+
+	return preImage, h.CloneBytes(), nil
```

Comment:
Shouldn't we just slice the hash directly instead of using `CloneBytes`.  i.e. `return preImage, h[:], nil`
Same Q above in the `SigHashSingle` conditional